### ISSUE TITLE
[Bug] Shared `InferenceSession.Run()` across threads crashes on every ult YOLO ONNX export (ORT CUDA-EP `Softmax` race) — MRE + bisect

### ISSUE THREAD

**Issue by narugo1992** [ultralytics/ultralytics#24520]
> Per @glenn-jocher's request in #14891 and #14969, opening this as a standalone issue with a self-contained MRE, `yolo checks` environment, and real text logs. The two closed threads pointed at the same root cause but were off-topic for those tickets.

### Ultralytics YOLO Component

Export, Predict

### Bug

Sharing one `onnxruntime.InferenceSession` (built with `CUDAExecutionProvider`) across N Python worker threads and calling `Run()` concurrently — **without a lock** — crashes or hangs on every ult-exported YOLO ONNX I've tested (v3 / v5 / v8 / v9 / v10 / v11 / v12, all sizes).

Symptoms vary by hardware but all trace to the same root cause:

* **GTX 1660 Ti** (compute 7.5): `CUDA failure 700: an illegal memory access was encountered` at `gpu_data_transfer.cc:cudaMemcpyAsync` after roughly 400–1 000 inferences with ≥ 8 worker threads
* **H200 / data-center GPUs**: same race manifests either as `CUDNN_STATUS_INTERNAL_ERROR` at `cudnnDestroy(cudnn_handle_)` (fast crash, ~10 s in) or a silent ~600 s deadlock, both triggering deterministically at ≥ 32 worker threads

Layer-by-layer bisect of `ultralytics.nn.modules.head.Detect.forward` (monkey-patch away `_inference()` first to compare against a raw 3-output export, then re-add each downstream op one at a time) localises the trigger to a **single operator: `Softmax(axis=...)`**. The ult export pipeline always emits this node for DFL decode; v11 C3k2 attention adds 1–2 more `Softmax(axis=-1)`, v12 area-attention adds 8 more. ORT's CUDA-EP dispatches `Softmax` to a cuDNN softmax whose internal handle is shared across the session's worker threads — that's the race window.

Replacing only that single `Softmax` node (V2b in the bisect) with a 5-op manual decomposition removes the symptom; every other op (`Concat`, `Reshape`, `Split`, `Mul`, `ReduceSum`, `Sigmoid` in the same neighbourhood) is concurrency-safe.

### Environment

```
Ultralytics 8.3.105 🚀 Python-3.10.20 torch-2.12.0+cu130 CUDA:0 (NVIDIA H200)
OS                  Linux-6.17.0-1008-gcp-x86_64-with-glibc2.39
Python              3.10.20
torch               2.12.0
torchvision         0.27.0
onnxruntime-gpu     1.23.2
onnx                1.21.0
NVIDIA driver       580.126.09
CUDA in runtime     12 (via pip-shipped nvidia-cuda-runtime-cu12 wheels)
GPU                 NVIDIA H200, 144 GB
```

Also reproduced on GTX 1660 Ti, 6 GB / driver 535 / CUDA 12, ORT 1.23.2, ult 8.3.105 — only the worker-count threshold differs.

### Minimal Reproducible Example

Self-contained, ~30 lines; deps are just `ultralytics`, `onnxruntime-gpu`, `numpy`. Tunable via `WORKERS` / `ROUNDS`.

```python
import threading, time, numpy as np, onnxruntime as ort
from ultralytics import YOLO

# 1. Export a fresh ult YOLO ONNX (the bug reproduces on every family/size)
YOLO("yolo11n.pt").export(format="onnx", imgsz=640, nms=False, simplify=True)

# 2. Build ONE shared InferenceSession on CUDA EP
so = ort.SessionOptions(); so.log_severity_level = 3
sess = ort.InferenceSession(
    "yolo11n.onnx",
    sess_options=so,
    providers=[("CUDAExecutionProvider", {"device_id": 0}), "CPUExecutionProvider"])

# 3. Hammer it from many threads, NO lock
WORKERS, ROUNDS = 32, 100
x = np.random.randn(1, 3, 640, 640).astype(np.float32)
errors = []

def worker():
    try:
        for _ in range(ROUNDS):
            sess.run(None, {"images": x})
    except BaseException as e:
        errors.append(f"{type(e).__name__}: {str(e)[:200]}")

threads = [threading.Thread(target=worker) for _ in range(WORKERS)]
t0 = time.perf_counter()
for t in threads: t.start()
for t in threads: t.join(timeout=600)
print(f"{len(errors)}/{WORKERS} workers errored | wall={time.perf_counter()-t0:.1f}s")
if errors:
    print("first error:", errors[0])
```

Expected on H200: `32/32 workers errored` within ~10 s with a `CUDNN_STATUS_INTERNAL_ERROR` at `cudnnDestroy`, or a 600 s `join` timeout if the race hits the deadlock path.
Drop `WORKERS` to 8 to reproduce on GTX 1660 Ti.

### Crash log (verbatim, from ORT)

Real trace captured from one of the runs (host name redacted to `<host>`):

```
[E:onnxruntime:Default, cuda_call.cc:123 CudaCall] CUDNN failure 4000: CUDNN_STATUS_INTERNAL_ERROR ;
  GPU=0 ; hostname=<host> ;
  file=/onnxruntime_src/onnxruntime/core/providers/cuda/cuda_execution_provider.cc ;
  line=191 ; expr=cudnnDestroy(cudnn_handle_);
```

On smaller cards the trace is instead:

```
[E:onnxruntime:Default, cuda_call.cc:123 CudaCall] CUDA failure 700: an illegal memory access
  was encountered ; GPU=0 ; file=.../gpu_data_transfer.cc ; expr=cudaMemcpyAsync(...)
```

Same race, different observable.

### What we did, briefly

Once the bisect pointed at `Softmax`, the operator was substituted in-place inside the ONNX graph with the math-equivalent

```
Y_max  = ReduceMax(X,    axes=[a], keepdims=1)
Y_shft = Sub(X, Y_max)
Y_exp  = Exp(Y_shft)
Y_sum  = ReduceSum(Y_exp, axes=[a], keepdims=1)
Y      = Div(Y_exp, Y_sum)
```

This is a graph-level rewrite of the exported `.onnx` file (no ORT or ult code change, no session config change, no env var). After the substitution, ORT's CUDA-EP no longer dispatches into cuDNN softmax for that network, and the race window is gone.

Validation we ran on this approach, summarised:

* **Concurrent stress** across all 33 ult-pretrained YOLO ONNX (v3 / v5 / v8 / v9 / v10 / v11 / v12, every size), 128 workers × 5 000 rounds × 33 models = **21.12 M attempted concurrent inferences on H200**. Aggregate **99.697 % succeed** (32 / 33 models perfect; yolov9e — the largest network — at 99.92 %, one of 128 workers in one of 10 chunks erroring).
* **Accuracy** on COCO128 via `ultralytics.val()`: 28 / 33 measurable models within **±0.001 mAP50-95** of the unpatched ONNX; max abs Δ = **0.000210**. (5 / 33 — yolov10s/m/b/l/x — hit a separate pre-existing `ultralytics.val()` postprocessing path bug affecting orig and patched symmetrically, so the comparison can't be drawn for those.)
* **Latency** (200 in-process calls per model, GPU EP, no concurrency): median p50 overhead **+0.035 ms**, mean +0.205 ms, max +2.04 ms (yolo12x — 9 softmax replacements × 5 ops = 45 extra kernel launches). 13 / 33 are zero-or-negative overhead.
* Cross-checked **bit-exact mAP50-95** (16-decimal-place match) on 4 production anime-detection ONNX models from `deepghs/anime_face_detection` and `deepghs/anime_head_detection` against their respective labelled validation sets.

Full methodology, per-model tables, figures and raw CSVs live in the technical report — too long to inline here:

* Tech report (root cause, validation, reproducer recipes): **[TECH_REPORT.md](https://github.com/deepghs/yolo-onnx-thread-safe/blob/main/TECH_REPORT.md)**
* Single-file patcher + `stress_test.py` harness: **[deepghs/yolo-onnx-thread-safe](https://github.com/deepghs/yolo-onnx-thread-safe)** (MIT-licensed)

### What this issue is / isn't asking for

It **is** a public record of: the symptom class, the bisect that pins it to one operator, the deterministic MRE, real text logs, and one externally-shipped graph-level workaround that's been validated end-to-end.

It **isn't** asking the ult team to vendor / link / change anything. The current [thread-safe inference guide](https://docs.ultralytics.com/guides/yolo-thread-safe-inference/) (one `InferenceSession` per thread) is a correct workaround for the same race; the external patcher above is just an alternative for users whose throughput budget makes per-thread session contention painful (we measured 97 imgs/s for the per-thread recipe vs 428 imgs/s for shared-session-no-lock on the patched ONNX at 32 workers / H200 / yolov8n).

If at some point it's useful to link from the guide, MIT licence makes that trivial — otherwise feel free to close this once it's been triaged; the goal was just to give the underlying bug a permanent searchable home with an MRE attached, per the request.

**Comment by UltralyticsAssistant** [comment]
👋 Hello @narugo1992, thank you for the detailed bug report and for providing a clear **minimum reproducible example**, environment details, and logs 🚀. This is an **automated response** to help with triage, and an Ultralytics engineer will also assist you soon.

We recommend a quick visit to the [Docs](https://docs.ultralytics.com/) for related guidance, including [Python](https://docs.ultralytics.com/usage/python) and [CLI](https://docs.ultralytics.com/usage/cli) usage examples.

Since this is a 🐛 Bug Report and you have already included an MRE, that is very helpful. If you are able to share any additional confirmation after testing on the latest package version and verified environments below, that can help the team reproduce and triage faster.

If this is a custom training ❓ Question, please provide as much information as possible, including dataset image examples and training logs, and verify you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips).

Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://www.reddit.com/r/Ultralytics/) to share knowledge with the community.

## Upgrade

Please upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) to verify this is not already resolved in the latest version:

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo26"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes) and [Tasks](https://docs.ultralytics.com/tasks) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

Thanks again for packaging this report so thoroughly 🧪. The MRE, environment output, and logs should make review much easier for the team.

**Comment by glenn-jocher** [comment]
Thanks for opening this as a standalone report with an MRE and text logs; for now the supported workaround remains the [thread-safe inference guide](https://docs.ultralytics.com/guides/yolo-thread-safe-inference) pattern of using one model or `InferenceSession` per thread, or serializing shared `sess.run()` calls with a `threading.Lock`, and we’ll keep this open for visibility since your bisect points to CUDA EP `Softmax` behavior rather than a model-family-specific export path.

**Comment by raimbekovm** [comment]
@onuralpszr Could you take a look please?

**Comment by onuralpszr** [comment]
> @onuralpszr Could you take a look please?

I will check it out sure

**Comment by onuralpszr** [comment]
@narugo1992  "Ultralytics 8.3.105 🚀 Python-3.10.20 torch-2.12.0+cu130 CUDA:0 (NVIDIA H200)" this is a very old package version did you tested with new package as well ?

**Comment by narugo1992** [comment]
@onuralpszr Yes — re-tested today on **ult 8.4.51** with onnxruntime-gpu 1.23.2 on NVIDIA H200 (same env spec as the original report). Fresh `.export(format="onnx", imgsz=640, nms=False)` from the new package still races.

Side-by-side, both exports carry exactly **one** `Softmax` (the DFL one, `axis=1`). Opset 18 → 20 changes the ops around it (extra `Reshape`, `Slice` instead of `Split`) but the cuDNN-handle race in ORT's CUDA EP doesn't go away — if anything it triggers _more_ easily:

| ult version | min config that hits 3/3 crashes (rounds=1000) | crash mode |
|---|---|---|
| 8.3.105 | 12 workers (incl. one ~30 s deadlock) | `CUDA failure 700: illegal memory access` |
| 8.4.51  | **8 workers** (all 3/3 hit `terminate called` core dump) | same `CUDA failure 700` |

Quick correction on my original MRE: the "≥ 32 workers" threshold was conservative from earlier GTX 1660 Ti runs. On H200 the floor is much lower — we measured 4w × 1000r ≈ 33 % trigger rate, 8w × 1000r 67–100 %, 16w × 1000r 100 % across 5 trials. The race is probabilistic by nature, so each config has its own trigger rate rather than a hard cutoff.

The graph-level `Softmax → ReduceMax + Sub + Exp + ReduceSum + Div` rewrite from the [tech report](https://github.com/deepghs/yolo-onnx-thread-safe/blob/main/TECH_REPORT.md) eliminates the race on **both** exports: 1.12 M concurrent inferences across 16/32/64/128 workers × 1000 rounds, zero failures, on both 8.3.105 and 8.4.51 ONNX. So this isn't being papered over by the 8.4 export path.

Happy to sweep additional ORT versions or other configs if useful for triage.
